mir/src/passes/unrolling/match_optimizer.rs

```diff
@@ -0,0 +1,241 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    ops::Deref,
+};
+
+use miden_diagnostics::SourceSpan;
+
+use crate::{
+    CompileError,
+    ir::{
+        Add, ConstantValue, Link, MatchArm, MirValue, Mul, Op, Parent, QuadFelt, RandomInputs,
+        SpannedMirValue, Sub, Value,
+    },
+    passes::duplicate_node,
+};
+
+type ConstraintEvaluationMap = BTreeMap<usize, Vec<(Link<Op>, Link<Op>)>>;
```

Comment:
Could we add a doc comment here explaining what the tuples in the vector represent? My understanding is that it is `(condition, constraint)` - but may be good to give an example.